test/integration/controller/controller_test.go

```diff
@@ -0,0 +1,721 @@
+				require.NotEmpty(t, diskName)
+				require.NoError(t, err)
+
+				_, err = crdProvisioner.PublishVolume(context.Background(), volumeID, "integration-test-node-0", nil, false, nil, nil)
+				require.NoError(t, err)
+
+				klog.Infof("Waiting for replica AzVolumeAttachment to be created for AzVolume (%s)", volumeName)
+				conditionFunc := func() (bool, error) {
+					labelSelector := fmt.Sprintf("%s=%s", azureutils.VolumeNameLabel, diskName)
+					azVolumeAttachments, err := azVolumeClient.DiskV1alpha1().AzVolumeAttachments(namespace).List(context.Background(), metav1.ListOptions{LabelSelector: labelSelector})
+					if err != nil {
+						return false, nil
+					}
+
+					for _, azVolumeAttachment := range azVolumeAttachments.Items {
+						if azVolumeAttachment.Spec.RequestedRole == v1alpha1.ReplicaRole && azVolumeAttachment.Status.Detail != nil {
+							klog.Infof("Replica AzVolumeAttachment (%s) successfully created and attached.", azVolumeAttachment.Name)
+							return true, nil
+						}
+					}
+					return false, nil
+				}
+
+				err = wait.PollImmediate(interval, timeout, conditionFunc)
+				require.NoError(t, err)
+
+				// delete volume
+				klog.Infof("Waiting for all AzVolumeAttachments attached to AzVolume (%s) to be detached and deleted", volumeName)
+				conditionFunc = func() (bool, error) {
+					err = crdProvisioner.DeleteVolume(context.Background(), volumeID, map[string]string{})
+					if err != nil {
+						return false, nil
+					}
+					return true, nil
+				}
+
+				err = wait.PollImmediate(interval, timeout, conditionFunc)
+				require.NoError(t, err)
+
+				// check if all azVolumeAttachments attached have been deleted
+				conditionFunc = func() (bool, error) {
+					labelSelector := fmt.Sprintf("%s=%s", azureutils.VolumeNameLabel, diskName)
+					azVolumeAttachments, err := azVolumeClient.DiskV1alpha1().AzVolumeAttachments(namespace).List(context.Background(), metav1.ListOptions{LabelSelector: labelSelector})
+					klog.Infof("Number of AzVolumeAttachment CRI attached to %s: %d", volumeName, len(azVolumeAttachments.Items))
+					if err != nil && !errors.IsNotFound(err) {
+						return false, err
+					}
+					return len(azVolumeAttachments.Items) == 0, nil
+				}
+
+				err = wait.PollImmediate(interval, timeout, conditionFunc)
+				require.NoError(t, err)
+			},
+			cleanUpFunc: nil,
+		},
+	}
+
+	for i, test := range tests {
+		t.Run(test.testName, func(t *testing.T) {
+			klog.Infof("AzVolume Test %d: %s", i, test.testName)
+
+			var err error
+			for retry := 0; retry < maxRetry; retry++ {
+				_, err = crdProvisioner.CreateVolume(context.Background(), test.volumeName, test.capacityRange, test.volumeCapabilities, test.parameters, test.secrets, test.volumeContentSource, test.accessibilityReq)
+				if err == nil {
+					break
+				}
+			}
+			require.NoError(t, err)
+
+			if test.testFunc != nil {
+				test.testFunc(t, test.volumeName)
+			}
+			if test.cleanUpFunc != nil {
```

Comment:
The `cleanUpFunc` will not be run if the test fails since you're calling it within the context of the `t.Run()` function. Is this intentional, or should the `cleanUpFunc` be run unconditionally?

Reply:
Thinking twice, clean up should run whether the main run fails or not to ensure that created resources get all deleted, so it should be in a different run session :)

Reply:
You could also use `defer` from within the same `t.Run()` session, e.g.
```
if test.cleanUpFunc != nil {
    defer test.CleanUpFunc(t)
}
if test.testFunc != nil {
    test.testFunc(t, test.volumeName)
}
```

Reply:
Ah good point